File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/component/StopListRow.kt
package com.mbta.tid.mbta_app.android.component

import android.content.Context
import androidx.compose.foundation.Image
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.pluralStringResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.mbta.tid.mbta_app.android.R
import com.mbta.tid.mbta_app.android.stopDetails.AlertCard
import com.mbta.tid.mbta_app.android.stopDetails.AlertCardSpec
import com.mbta.tid.mbta_app.android.stopDetails.StopDot
import com.mbta.tid.mbta_app.android.stopDetails.TripRouteAccents
import com.mbta.tid.mbta_app.android.util.SettingsCache
import com.mbta.tid.mbta_app.android.util.Typography
import com.mbta.tid.mbta_app.android.util.modifiers.DestinationPredictionBalance
import com.mbta.tid.mbta_app.android.util.modifiers.placeholderIfLoading
import com.mbta.tid.mbta_app.android.util.routeModeLabel
import com.mbta.tid.mbta_app.model.Alert
import com.mbta.tid.mbta_app.model.AlertSummary
import com.mbta.tid.mbta_app.model.Route
import com.mbta.tid.mbta_app.model.RouteBranchSegment
import com.mbta.tid.mbta_app.model.SegmentAlertState
import com.mbta.tid.mbta_app.model.Stop
import com.mbta.tid.mbta_app.model.UpcomingFormat
import com.mbta.tid.mbta_app.repositories.Settings
import kotlinx.serialization.Serializable

class StopPlacement(val isFirst: Boolean = false, val isLast: Boolean = false)

@Serializable
sealed class StopListContext {
    data object Trip : StopListContext()

    data object RouteDetails : StopListContext()
}

@Composable
fun StopListRow(
    stop: Stop,
    stopLane: RouteBranchSegment.Lane,
    stickConnections: List<RouteBranchSegment.StickConnection>,
    onClick: () -> Unit,
    routeAccents: TripRouteAccents,
    stopListContext: StopListContext,
    modifier: Modifier = Modifier,
    activeElevatorAlerts: Int = 0,
    alertSummaries: Map<String, AlertSummary?> = emptyMap(),
    connectingRoutes: List<Route>? = null,
    disruption: UpcomingFormat.Disruption? = null,
    getAlertState: (fromStop: String, toStop: String) -> SegmentAlertState = { _, _ ->
        SegmentAlertState.Normal
    },
    onClickLabel: String? = null,
    stopPlacement: StopPlacement = StopPlacement(),
    onOpenAlertDetails: (Alert) -> Unit = {},
    targeted: Boolean = false,
    trackNumber: String? = null,
    descriptor: @Composable () -> Unit = {},
    rightSideContent: @Composable RowScope.(Modifier) -> Unit = {},
) {
    val context = LocalContext.current
    val showStationAccessibility = SettingsCache.get(Settings.StationAccessibility)

    Column {
        Box(
            Modifier.padding(
                    horizontal = if (stopListContext is StopListContext.Trip) 6.dp else 0.dp
                )
                .then(modifier)
                .height(IntrinsicSize.Min)
                .defaultMinSize(minHeight = 48.dp)
        ) {
            if (!stopPlacement.isLast && !targeted && disruption == null) {
                HaloSeparator(Modifier.align(Alignment.BottomCenter))
            }
            Row(
                Modifier.fillMaxHeight()
                    .padding(start = 8.dp)
                    .semantics(mergeDescendants = true) {}
                    .clickable(onClickLabel = onClickLabel) { onClick() },
                horizontalArrangement = Arrangement.spacedBy(0.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                RouteLine(routeAccents, stopLane, stickConnections, targeted, getAlertState)
                Column(Modifier.padding(vertical = 12.dp).padding(start = 8.dp)) {
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                    ) {
                        if (
                            showStationAccessibility &&
                                (activeElevatorAlerts > 0 || !stop.isWheelchairAccessible)
                        ) {
                            Row(
                                Modifier.padding(start = 6.dp).widthIn(max = 28.dp),
                                verticalAlignment = Alignment.CenterVertically,
                                horizontalArrangement = Arrangement.Center,
                            ) {
                                if (activeElevatorAlerts > 0) {
                                    Image(
                                        modifier = Modifier.height(18.dp).testTag("elevator_alert"),
                                        painter =
                                            painterResource(R.drawable.accessibility_icon_alert),
                                        contentDescription = null,
                                    )
                                } else {
                                    Image(
                                        modifier =
                                            Modifier.height(18.dp)
                                                .testTag("wheelchair_not_accessible")
                                                .placeholderIfLoading(),
                                        painter =
                                            painterResource(
                                                R.drawable.accessibility_icon_not_accessible
                                            ),
                                        contentDescription = null,
                                    )
                                }
                            }
                        }
                        Column(
                            DestinationPredictionBalance.destinationWidth(),
                            horizontalAlignment = Alignment.Start,
                            verticalArrangement = Arrangement.spacedBy(4.dp),
                        ) {
                            descriptor()
                            Text(
                                stop.name,
                                Modifier.semantics {
                                        contentDescription =
                                            stopAccessibilityLabel(
                                                stop,
                                                targeted,
                                                stopPlacement.isFirst,
                                                context,
                                            )
                                    }
                                    .placeholderIfLoading(),
                                color = colorResource(R.color.text),
                                style = if (targeted) Typography.headlineBold else Typography.body,
                            )
                            if (trackNumber != null) {
                                Text(
                                    stringResource(R.string.track_number, trackNumber),
                                    Modifier.semantics {
                                            contentDescription =
                                                context.getString(
                                                    R.string.boarding_track,
                                                    trackNumber,
                                                )
                                        }
                                        .placeholderIfLoading(),
                                    color = MaterialTheme.colorScheme.onPrimary,
                                    style = Typography.footnote,
                                )
                            }
                        }
                        rightSideContent(Modifier.padding(end = 12.dp))
                        // Adding the accessibility description into the stop label rather than on
                        // the accessibility icon so that it is clear which stop it is associated
                        // with. Empty Rows, Canvas, Text that don't take up size, etc. do not have
                        // their content descriptions read.
                        val stopNotAccessibleContentDescription =
                            stringResource(R.string.not_accessible_stop_card)
                        val elevatorAlertContentDescription =
                            pluralStringResource(
                                R.plurals.elevator_closure_count_accessibility_description,
                                activeElevatorAlerts,
                                activeElevatorAlerts,
                            )

                        Text(
                            "",
                            modifier =
                                Modifier.semantics {
                                        contentDescription =
                                            if (
                                                showStationAccessibility &&
                                                    !stop.isWheelchairAccessible
                                            ) {
                                                stopNotAccessibleContentDescription
                                            } else if (
                                                showStationAccessibility && activeElevatorAlerts > 0
                                            ) {
                                                elevatorAlertContentDescription
                                            } else {
                                                ""
                                            }
                                    }
                                    .width(1.dp)
                                    .height(1.dp),
                        )
                    }

                    if (!connectingRoutes.isNullOrEmpty()) {
                        ScrollRoutes(
                            connectingRoutes,
                            Modifier.clearAndSetSemantics {
                                contentDescription =
                                    scrollRoutesAccessibilityLabel(connectingRoutes, context)
                            },
                        )
                    }
                }
            }
        }
        if (disruption != null) {
            Box(Modifier.height(IntrinsicSize.Min)) {
                AlertCard(
                    disruption.alert,
                    alertSummaries[disruption.alert.id],
                    AlertCardSpec.Downstream,
                    routeAccents.color,
                    routeAccents.textColor,
                    onViewDetails = { onOpenAlertDetails(disruption.alert) },
                    interiorPadding = PaddingValues(start = 10.dp),
                )
            }
        }
    }
}

private fun connectionLabel(route: Route, context: Context) = routeModeLabel(context, route)

@Composable
fun ScrollRoutes(routes: List<Route>, modifier: Modifier = Modifier) {
    Row(
        modifier.horizontalScroll(rememberScrollState()).padding(top = 8.dp, end = 20.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        for (route in routes) {
            RoutePill(route, type = RoutePillType.Flex)
        }
    }
}

// leading "." creates sentence break
private fun scrollRoutesAccessibilityLabel(
    connectingRoutes: List<Route>,
    context: Context,
): String =
    ". " +
        when {
            connectingRoutes.isEmpty() -> ""
            connectingRoutes.size == 1 ->
                context.getString(
                    R.string.connection_to,
                    connectionLabel(connectingRoutes.single(), context),
                )
            else -> {
                val firstConnections = connectingRoutes.dropLast(1)
                val lastConnection = connectingRoutes.last()
                context.getString(
                    R.string.connections_to_and,
                    firstConnections.joinToString(separator = ", ") {
                        connectionLabel(it, context)
                    },
                    connectionLabel(lastConnection, context),
                )
            }
        }

private fun stopAccessibilityLabel(
    stop: Stop,
    targeted: Boolean,
    firstStop: Boolean,
    context: Context,
): String {
    val name = stop.name
    return when {
        targeted && firstStop -> context.getString(R.string.selected_stop_first_stop, name)
        targeted -> context.getString(R.string.selected_stop, name)
        firstStop -> context.getString(R.string.first_stop, name)
        else -> name
    }
}

@Composable
private fun RouteLine(
    routeAccents: TripRouteAccents,
    stopLane: RouteBranchSegment.Lane,
    stickConnections: List<RouteBranchSegment.StickConnection>,
    targeted: Boolean,
    getAlertState: (fromStop: String, toStop: String) -> SegmentAlertState,
) {
    Box(contentAlignment = Alignment.Center, modifier = Modifier.fillMaxHeight().width(40.dp)) {
        Column(Modifier.fillMaxHeight()) {
            StickDiagram(
                routeAccents.color,
                stickConnections,
                Modifier.weight(1f),
                getAlertState = getAlertState,
            )
        }
        StopDot(
            routeAccents,
            targeted,
            when (stopLane) {
                RouteBranchSegment.Lane.Left -> Modifier.padding(end = 20.dp)
                RouteBranchSegment.Lane.Center -> Modifier
                RouteBranchSegment.Lane.Right -> Modifier.padding(start = 20.dp)
            },
        )
    }
}

File: iosApp/iosApp/ComponentViews/StopListRow.swift
//
//  StopListRow.swift
//  iosApp
//
//  Created by Melody Horn on 7/8/25.
//  Copyright © 2025 MBTA. All rights reserved.
//

import Shared
import SwiftUI

struct StopPlacement {
    let isFirst: Bool
    let isLast: Bool

    init(isFirst: Bool = false, isLast: Bool = false) {
        self.isFirst = isFirst
        self.isLast = isLast
    }
}

enum StopListContext {
    case trip
    case routeDetails
}

struct StopListRow<Descriptor: View, RightSideContent: View>: View {
    var stop: Stop
    var stopLane: RouteBranchSegment.Lane
    var stickConnections: [RouteBranchSegment.StickConnection]
    var onClick: () -> Void
    var routeAccents: TripRouteAccents
    var stopListContext: StopListContext
    var activeElevatorAlerts: Int
    var alertSummaries: [String: AlertSummary?]
    var background: Color?
    var connectingRoutes: [Route]?
    var disruption: UpcomingFormat.Disruption?
    var getAlertState: (_ fromStop: String, _ toStop: String) -> SegmentAlertState
    var stopPlacement: StopPlacement
    var onOpenAlertDetails: (Shared.Alert) -> Void
    var targeted: Bool
    var trackNumber: String?
    var descriptor: () -> Descriptor
    var rightSideContent: () -> RightSideContent

    @EnvironmentObject var settingsCache: SettingsCache
    var showStationAccessibility: Bool { settingsCache.get(.stationAccessibility) }

    init(
        stop: Stop,
        stopLane: RouteBranchSegment.Lane,
        stickConnections: [RouteBranchSegment.StickConnection],
        onClick: @escaping () -> Void,
        routeAccents: TripRouteAccents,
        stopListContext: StopListContext,
        activeElevatorAlerts: Int = 0,
        alertSummaries: [String: AlertSummary?] = [:],
        background: Color? = nil,
        connectingRoutes: [Route]? = nil,
        disruption: UpcomingFormat.Disruption? = nil,
        getAlertState: @escaping (_ fromStop: String, _ toStop: String) -> SegmentAlertState = { _, _ in .normal },
        stopPlacement: StopPlacement = .init(),
        onOpenAlertDetails: @escaping (Shared.Alert) -> Void = { _ in },
        targeted: Bool = false,
        trackNumber: String? = nil,
        descriptor: @escaping () -> Descriptor,
        rightSideContent: @escaping () -> RightSideContent
    ) {
        self.stop = stop
        self.stopLane = stopLane
        self.stickConnections = stickConnections
        self.onClick = onClick
        self.routeAccents = routeAccents
        self.stopListContext = stopListContext
        self.activeElevatorAlerts = activeElevatorAlerts
        self.alertSummaries = alertSummaries
        self.background = background
        self.connectingRoutes = connectingRoutes
        self.disruption = disruption
        self.getAlertState = getAlertState
        self.stopPlacement = stopPlacement
        self.onOpenAlertDetails = onOpenAlertDetails
        self.targeted = targeted
        self.trackNumber = trackNumber
        self.descriptor = descriptor
        self.rightSideContent = rightSideContent
    }

    var body: some View {
        VStack(spacing: 0) {
            stopRow
                .background(background)
                .overlay {
                    if targeted {
                        VStack {
                            HaloSeparator(height: 1)
                            Spacer()
                            HaloSeparator(height: 2)
                        }
                    }
                }
            if let disruption {
                ZStack(alignment: .leading) {
                    AlertCard(
                        alert: disruption.alert,
                        alertSummary: alertSummaries[disruption.alert.id] ?? nil,
                        spec: .downstream,
                        color: routeAccents.color,
                        textColor: routeAccents.textColor,
                        onViewDetails: { onOpenAlertDetails(disruption.alert) },
                        internalPadding: .init(top: 0, leading: 5, bottom: 0, trailing: 0)
                    )
                    .padding(.horizontal, -4)
                }
            }
        }
        .fixedSize(horizontal: false, vertical: true).padding(.horizontal, stopListContext == .trip ? 7 : 0)
    }

    @ViewBuilder
    var stopRow: some View {
        ZStack(alignment: .bottom) {
            if !stopPlacement.isLast, !targeted, disruption == nil {
                HaloSeparator()
            }
            HStack(alignment: .center, spacing: 0) {
                routeLine
                VStack(alignment: .leading, spacing: 8) {
                    Button(
                        action: { onClick() },
                        label: {
                            HStack(alignment: .center, spacing: 0) {
                                if showStationAccessibility, activeElevatorAlerts > 0 || !stop.isWheelchairAccessible {
                                    HStack(alignment: .center) {
                                        if showStationAccessibility, activeElevatorAlerts > 0 {
                                            Image(.accessibilityIconAlert).tag("elevator_alert")
                                        } else {
                                            Image(.accessibilityIconNotAccessible).tag("wheelchair_not_accessible")
                                        }
                                    }
                                    .accessibilityHidden(true)
                                    .frame(minWidth: 28, maxWidth: 28)
                                    .padding(.horizontal, 6)
                                }
                                VStack(alignment: .leading, spacing: 4) {
                                    descriptor()
                                    Text(stop.name)
                                        .font(targeted ? Typography.headlineBold : Typography.body)
                                        .foregroundStyle(Color.text)
                                        .multilineTextAlignment(.leading)
                                        .accessibilityLabel(stopAccessibilityLabel)
                                    if let trackNumber {
                                        Text("Track \(trackNumber)")
                                            .font(Typography.footnote)
                                            .foregroundStyle(Color.text)
                                            .multilineTextAlignment(.leading)
                                            .accessibilityLabel(Text("Boarding on track \(trackNumber)"))
                                    }
                                }
                                Spacer()
                                rightSideContent()

                                // Adding the accessibility description into the stop label rather than on the
                                // accessibility icon so that it is clear which stop it is associated with
                                if showStationAccessibility, activeElevatorAlerts > 0 {
                                    HStack {}
                                        .accessibilityLabel(Text(
                                            "This stop has \(activeElevatorAlerts, specifier: "%ld") elevators closed",
                                            comment: "Describe an elevator outage at the stop in the list of all stops on the trip"
                                        ))
                                } else if showStationAccessibility, !stop.isWheelchairAccessible {
                                    HStack {}
                                        .accessibilityLabel(Text("This stop is not accessible"))
                                }
                            }
                        }
                    )
                    .preventScrollTaps()
                    .accessibilityElement(children: .combine)
                    .accessibilityInputLabels([stop.name])
                    .accessibilityAddTraits(.isHeader)
                    .accessibilityHeading(.h4)

                    if let connectingRoutes, !connectingRoutes.isEmpty {
                        scrollRoutes
                            .accessibilityElement()
                            .accessibilityLabel(scrollRoutesAccessibilityLabel)
                    }
                }
                .accessibilitySortPriority(1)
                .padding(.leading, 8)
                .padding(.vertical, 12)
                .padding(.trailing, 8)
                .padding(.bottom, stopPlacement.isLast ? 0 : 1)
                .frame(minHeight: 56)
            }.accessibilityElement(children: .contain)
        }
    }

    @ViewBuilder
    var routeLine: some View {
        ZStack(alignment: .center) {
            VStack(spacing: 0) {
                StickDiagram(routeAccents.color, stickConnections, getAlertState: getAlertState)
            }
            let dot = StopDot(routeAccents: routeAccents, targeted: targeted)
            switch stopLane {
            case .left: dot.padding(.trailing, 20)
            case .center: dot.padding(.horizontal, 10)
            case .right: dot.padding(.leading, 20)
            }
        }
        .padding(.horizontal, 7)
    }

    func connectionLabel(route: Route) -> String { routeModeLabel(route: route) }

    var scrollRoutes: some View {
        let routeView = ScrollView(.horizontal, showsIndicators: false) {
            HStack {
                if let connectingRoutes {
                    ForEach(connectingRoutes, id: \.id) { route in
                        RoutePill(route: route, line: nil, type: .flex)
                    }
                }
            }.padding(.trailing, 20)
        }.onTapGesture { onClick() }
        return routeView.scrollBounceBehavior(.basedOnSize, axes: [.horizontal])
    }

    var scrollRoutesAccessibilityLabel: String {
        guard let connectingRoutes, !connectingRoutes.isEmpty else { return "" }
        if connectingRoutes.count == 1 {
            return String(format: NSLocalizedString(
                "Connection to %@",
                comment: "VoiceOver label for a single connecting route at a stop, ex 'Connection to 1 bus'"
            ), connectionLabel(route: connectingRoutes.first!))
        } else {
            let firstConnections = connectingRoutes.prefix(connectingRoutes.count - 1)
            let lastConnection = connectingRoutes.last!
            return String(
                format: NSLocalizedString(
                    "Connections to %@ and %@",
                    comment: """
                    VoiceOver label for multiple connecting routes at a stop,
                    ex 'Connections to Red Line train, 71 bus, and 73 bus',
                    the first replaced value can be any number of comma separated route labels
                    """
                ),
                firstConnections.map(connectionLabel).joined(separator: ", "),
                connectionLabel(route: lastConnection)
            )
        }
    }

    var stopAccessibilityLabel: String {
        let name = stop.name
        return if targeted, stopPlacement.isFirst {
            String(format: NSLocalizedString(
                "%@, selected stop, first stop",
                comment: "Screen reader text for a stop name on the stop details page when that stop is both selected and first"
            ), name)
        } else if targeted {
            String(format: NSLocalizedString(
                "%@, selected stop",
                comment: "Screen reader text for a stop name on the stop details page when that stop is the selected one"
            ), name)
        } else if stopPlacement.isFirst {
            String(format: NSLocalizedString(
                "%@, first stop",
                comment: "Screen reader text for a stop name on the stop details page when that stop is the first stop on the line"
            ), name)
        } else {
            name
        }
    }
}
